### PR TITLE
Bugfix/#399 navigation

### PR DESCRIPTION
**Pull request by ziflex**
Big changes!
Refactored the way how we deal with frames and network events:
- new EventLoop (no hardcoded EventBroker anymore)
- new DOM manager
- new Network manager
- more accurate event handling
- faster Document update
- redirects handling

### CODE REVIEW

File: pkg/drivers/cdp/dom/helpers.go [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+	"github.com/PuerkitoBio/goquery"
+	"github.com/mafredri/cdp"
+	"github.com/mafredri/cdp/protocol/dom"
+	"github.com/mafredri/cdp/protocol/page"
+	"github.com/mafredri/cdp/protocol/runtime"
+)
+
+var emptyExpires = time.Time{}
+
+func parseAttrs(attrs []string) *values.Object {
```
Comment (3timeslazy) [inline pkg/drivers/cdp/dom/helpers.go]: I don’t quite understand what is going on here

Comment (ziflex) [reply pkg/drivers/cdp/dom/helpers.go]: CDP returns HTML attributes as an array of strings of key-value pairs. This methods parses the array and converts it into an object.

File: pkg/drivers/cdp/events/listeners.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+	}
+
+	bucket[listener.ID] = listener
+}
+
+func (lc *ListenerCollection) Remove(eventID ID, listenerID ListenerID) {
+	lc.mu.Lock()
+	defer lc.mu.Unlock()
+
+	bucket, exists := lc.values[eventID]
```
[the remark this answers is not in the document]
Comment (ziflex) [reply pkg/drivers/cdp/events/listeners.go]: Yes, but i'm not deleting it from ``lc.values``. I'm deleting in from a nested map.

File: pkg/drivers/cdp/events/source.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
+package events
+
+import (
+	"github.com/mafredri/cdp/rpcc"
+)
+
+type (
```
[the remark this answers is not in the document]
Comment (ziflex) [reply pkg/drivers/cdp/events/source.go]: :(

File: pkg/drivers/cdp/events/source.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+	GenericSource struct {
+		eventID ID
+		stream  rpcc.Stream
+		recv    func(stream rpcc.Stream) (interface{}, error)
+	}
+)
+
+var (
+	Error = New("error")
```
[the remark this answers is not in the document]
Comment (ziflex) [reply pkg/drivers/cdp/events/source.go]: It's a builtin event ID

File: pkg/drivers/cookies.go [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+}
+
+func (c HTTPCookies) Compare(other core.Value) int64 {
+	if other.Type() != HTTPCookiesType {
+		return Compare(HTTPCookiesType, other.Type())
+	}
+
+	oc := other.(HTTPCookies)
+
+	if len(c) > len(oc) {
```
[the remark this answers is not in the document]
Comment (ziflex) [reply pkg/drivers/cookies.go]: Maybe? I didn't know that you can omit switch expression.

**Pull request merged**